<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="Beetl : BEETL: Burrows-Wheeler Extended Tool Library" />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Beetl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/BEETL/BEETL">View on GitHub</a>

          <h1 id="project_title">Beetl</h1>
          <h2 id="project_tagline">BEETL: Burrows-Wheeler Extended Tool Library</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/BEETL/BEETL/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/BEETL/BEETL/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>BEETL: Burrows-Wheeler Extended Tool Library</h1>

<p>BEETL is a suite of applications for building and manipulating the 
Burrows-Wheeler Transform (BWT) of collections of DNA sequences. 
The algorithms employed in BEETL are intended to scale to collections of 
sequences containing one billion entries or more.</p>

<p>The initial release implemented two flavours of an algorithm for
building the BWT of a sequence collection - BCR and BCRext.</p>

<p>Subsequent releases are adding functionality for efficient inversion and
querying of BWTs.</p>

<p>Questions and queries can be sent to the BEETL <a href="http://tech.groups.yahoo.com/group/BEETL/">mailing list</a></p>

<h2>Release History</h2>

<h3>Version 0.10.0 (17th June 2014)</h3>
<ul>
<li>New beetl-fastq tool to convert FASTQ files into a set of smaller and searchable files (including quality scores and read ids)</li>
<li>Added RLE53 class (run length encoding: 3 bits for base, 5 bits for run length), but RLE44 is still our default BWT format</li>
<li>Beetl-convert handles BWT44 -> BWT53 conversions</li>
<li>Integrated LCP code with main BCR, making it faster</li>
<li>Beetl-extend can now rebuild the whole reads' bases with --propagate-sequence (instead of stopping at the $ signs)</li>
<li>Replaced compile-time PROPAGATE_SEQUENCE with runtime --propagate-sequence in beetl-compare</li>
<li>Beetl-compare output now correctly goes into the directory specified by --output</li>
<li>Beetl-search/beetl-extend --use-shm stores pre-parsed index files in shared memory for faster tool startup when doing many small searches</li>
<li>Thanks to <em>Shaun Jackman</em> and <em>Rayan Chikhi</em> for compilation patches</li>
</ul>

<h3>Version 0.9.0 (9th December 2013)</h3>
<ul>
<li>New K-mer search and propagation tools (see example with beetl-search/beetl-extend in doc/BEETL.md)</li>
<li>MetaBeetl filtering changed a bit, in preparation for further improvements</li>
<li>MetaBeetl database construction scripts and instructions updated</li>
<li>Better distribution packaging, with 'make distcheck'</li>
</ul>

<h3>Version 0.8.0 (29th October 2013)</h3>

<ul>
<li>New BWT-based tumour-normal filter (see example in doc/BEETL.md)</li>
<li>Beetl-compare modes: 'split' renamed to 'splice'</li>
<li>Beetl-compare modes: new 'tumour-normal'</li>
<li>16-way parallel beetl-compare</li>
<li>MetaBEETL is now able to run without sequence propagation (which is faster)</li>
<li>Upgraded requirement: autoconf 2.65 (improved regression testing)</li>
</ul>

<h3>Version 0.7.0 (18th September 2013)</h3>

<ul>
<li>New beetl-correct tool</li>
<li>Updated regression tests</li>
<li>Detection of required libraries and OpenMP in configure script</li>
</ul>

<h3>Version 0.6.0 (27th August 2013)</h3>

<ul>
<li><strong>Important</strong>: Harmonised file encodings - you may need to rebuild your BWT files or just convert bwt-B00 from BWT_ASCII to BWT_RLE with beetl-convert (bwt-B00 was always ASCII-encoded. It now follows the same encoding as the other piles)</li>
<li>Now requires GCC with C++11 support (gcc 4.3 or above)</li>
<li>New Markdown documentation style</li>
<li>Faster beetl-compare and metaBEETL</li>
<li>Lossless quality compression scripts added, even though they are not using BWT</li>
<li>New features:
<ul>
<li>BWT creation with reverse-complemented reads <em>(--add-rev-comp)</em></li>
<li>BWT creation with multiple reads per sequence <em>(--sequence-length)</em></li>
</ul>
</li>
<li>Support for BCL and BCL.gz in beetl-convert</li>
</ul>

<h3>Version 0.5.0 (10th June 2013)</h3>
Faster metaBEETL
<ul>
<li>Pruning of BKPT to reduce the number of MTAXA tests and output lines</li>
<li>New beetl-compare <em>--subset</em> option for distributed computing</li>
</ul>

<h3>Version 0.4.0 (10th April 2013)</h3>
Robustness improvements

<h3>Version 0.3.0 (7th April 2013)</h3>
Metagenomics <a href="http://dx.doi.org/10.1186/1471-2105-14-S5-S2">metaBEETL</a> code added to main tree

<h3>Version 0.2.0 (19th March 2013)</h3>
<ul>
<li><a href="http://arxiv.org/abs/1305.0160">Longest Common Prefix computation</a> integrated under <em>--generate-lcp</em> in beetl-bwt</li>
<li>New <em>beetl-unbwt</em> command line interface</li>
</ul>
<h3>Version 0.1.0 (28th February 2013)</h3>

<ul>
<li>New command line interface</li>
<li>Support for FASTA, FASTQ, raw SEQ and cycle-by-cycle files as input</li>
<li>Distinct intermediate and final formats</li>
<li>Support for Huffman encoding as intermediate or final format</li>
<li>Support for intermediate "incremental run-length-encoded" format</li>
<li>Automatic detection of format</li>
<li>Performance prediction for automatic use of fastest algorithm and options</li>
</ul>

<h3>Version 0.0.2 (25th June 2012)</h3>
Code to build SAP array, as described in <a href="http://arxiv.org/abs/1205.0192">this paper</a>.

<h3>Version 0.0.1 (18th November 2011)</h3>

This contains initial implementations of the BCR and BCRext algorithms
as described in <a href="http://dx.doi.org/10.1007/978-3-642-21458-5_20">our CPM paper</a>.


<h2>Contributors to BEETL, past and present</h2>

<ul>
<li>Christina Ander, University of Bielefeld</li>
<li>Markus J. Bauer, Illumina UK <em>(now at Boehringer Ingelheim)</em></li>
<li><a href="https://twitter.com/coxtonyj">Anthony J. Cox</a>, Illumina UK <em>(project lead)</em></li>
<li>Tobias Jakobi, University of Bielefeld</li>
<li>Lilian Janin, Illumina UK</li>
<li>Giovanna Rosone, University of Palermo</li>
<li>Ole Schulz-Trieglaff, Illumina UK</li>
</ul>

<h2>BEETL-ography: papers about BEETL</h2>

<p>
<li><strong>Adaptive reference-free compression of sequence quality scores</strong></li>
<li>Lilian Janin, Giovanna Rosone, Anthony J. Cox</li>
<li><em>Bioinformatics</em>, in press (2013)</li>
<li><a href="http://dx.doi.org/10.1093/bioinformatics/btt257">Link to paper</a></li>
<li><a href="http://arxiv.org/abs/1305.0159">Link to preprint</a></li>  
</p>

<p>
<li><strong>Lightweight algorithms for constructing and inverting the BWT of string collections</strong></li>
<li>Markus J Bauer, Anthony J Cox, Giovanna Rosone</li>
<li><em>Theoretical Computer Science</em>, Volume 483, 29 April 2013, Pages 134-148, ISSN 0304-3975</li> 
<li><a href="http://dx.doi.org/10.1016/j.tcs.2012.02.002">Link to paper</a></li>
</p>

<p>
<li><strong>metaBEETL: high-throughput analysis of heterogeneous microbial populations from shotgun DNA sequences</strong></li>
<li>Christina Ander, Ole B Schulz-Trieglaff, Jens Stoye and Anthony J Cox</li>
<li><em>BMC Bioinformatics 2013</em>, 14(Suppl 5):S2</li>
<li><a href="http://dx.doi.org/10.1186/1471-2105-14-S5-S2">Link to paper</a></li>
</p>

<p>
<li><strong>Lightweight LCP construction for next-generation sequencing datasets</strong></li>
<li>Markus J Bauer, Anthony J Cox, Giovanna Rosone, Marinella Sciortino</li>
<li>In <em>Algorithms in Bioinformatics (Lecture Notes in Computer Science 7534)</em>. Springer Berlin Heidelberg, 2012. 326-337.</li>
<li><a href="http://dx.doi.org/10.1007/978-3-642-33122-0_26">Link to paper</a></li>
<li><a href="http://arxiv.org/abs/1305.0160">Link to preprint</a></li>
</p>

<p>
<li><strong>Comparing DNA sequence collections by direct comparison of compressed text indexes</strong></li>
<li>Anthony J. Cox, Tobias Jakobi, Giovanna Rosone, Ole B. Schulz-Trieglaff</li>
<li>In <em>Algorithms in Bioinformatics (Lecture Notes in Computer Science 7534)</em>. Springer Berlin Heidelberg, 2012. 214-224.</li>
<li><a href="http://dx.doi.org/10.1007/978-3-642-33122-0_17">Link to paper</a></li>
<li><a href="http://arxiv.org/abs/1304.5535">Link to preprint</a></li>
</p>

<p>
<li><strong>Large-scale compression of genomic sequence databases with the Burrows-Wheeler transform</strong></li>
<li>Anthony J. Cox, Markus J. Bauer, Tobias Jakobi, Giovanna Rosone</li>
<li><em>Bioinformatics</em> (2012) 28 (11): 1415-1419. </li>
<li><a href="http://dx.doi.org/10.1093/bioinformatics/bts173">Link to paper</a></li> 
<li><a href="http://arxiv.org/abs/1205.0192">Link to preprint</a></li>
</p>

<p>
<li><strong>Lightweight BWT construction for very large string collections</strong></li>
<li>Markus J Bauer, Anthony J Cox, Giovanna Rosone</li>
<li>In <em>Combinatorial Pattern Matching (Lecture Notes in Computer Science 6661)</em>. Springer Berlin Heidelberg, 2011. 219-231.</li>
<li><a href="http://dx.doi.org/10.1007/978-3-642-21458-5_20">Link to paper</a></li>
</p>



      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Beetl maintained by <a href="https://github.com/BEETL">BEETL</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
